extension/socks5proxy/bin/windowsbypass.cpp

```diff
@@ -0,0 +1,376 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "windowsbypass.h"
+
+#include <WS2tcpip.h>
+#include <netioapi.h>
+#include <windows.h>
+#include <winsock2.h>
+
+#include <QAbstractSocket>
+#include <QHostAddress>
+#include <QScopeGuard>
+#include <QUuid>
+
+#include "socks5.h"
+
+// Fixed GUID of the Wireguard NT driver.
+constexpr const QUuid WIREGUARD_NT_GUID(0xf64063ab, 0xbfee, 0x4881, 0xbf, 0x79,
+                                        0x36, 0x6e, 0x4c, 0xc7, 0xba, 0x75);
+
+// Called by the kernel on network interface changes.
+// Runs in some unknown thread, so invoke a Qt signal to do the real work.
+static void netChangeCallback(PVOID context, PMIB_IPINTERFACE_ROW row,
+                              MIB_NOTIFICATION_TYPE type) {
+  WindowsBypass* bypass = static_cast<WindowsBypass*>(context);
+  Q_UNUSED(type);
+
+  // Invoke the route changed signal to do the real work in Qt.
+  QMetaObject::invokeMethod(bypass, "refreshIfMetrics", Qt::QueuedConnection);
+}
+
+// Called by the kernel on network interface changes.
+// Runs in some unknown thread, so invoke a Qt signal to do the real work.
+static void addrChangeCallback(PVOID context, PMIB_UNICASTIPADDRESS_ROW row,
+                               MIB_NOTIFICATION_TYPE type) {
+  WindowsBypass* bypass = static_cast<WindowsBypass*>(context);
+  Q_UNUSED(type);
+
+  // Invoke the route changed signal to do the real work in Qt.
+  QMetaObject::invokeMethod(bypass, "refreshAddresses", Qt::QueuedConnection);
+}
+
+// Called by the kernel on route changes.
+// Runs in some unknown thread, so invoke a Qt signal to do the real work.
+static void routeChangeCallback(PVOID context, PMIB_IPFORWARD_ROW2 row,
+                                MIB_NOTIFICATION_TYPE type) {
+  WindowsBypass* bypass = static_cast<WindowsBypass*>(context);
+  Q_UNUSED(type);
+
+  // Ignore routing changes into the Wireguard tunnel.
+  int family = AF_UNSPEC;
+  if (row) {
+    family = row->DestinationPrefix.Prefix.si_family;
+  }
+
+  // Invoke the route changed signal to do the real work in Qt.
+  QMetaObject::invokeMethod(bypass, "refreshRoutes", Qt::QueuedConnection,
+                            Q_ARG(int, family));
+}
+
+WindowsBypass::WindowsBypass(Socks5* proxy) : QObject(proxy) {
+  connect(proxy, &Socks5::outgoingConnection, this,
+          &WindowsBypass::outgoingConnection);
+
+  NotifyIpInterfaceChange(AF_UNSPEC, netChangeCallback, this, true,
+                          &m_netChangeHandle);
+  NotifyUnicastIpAddressChange(AF_UNSPEC, addrChangeCallback, this, true,
+                               &m_addrChangeHandle);
+  NotifyRouteChange2(AF_UNSPEC, routeChangeCallback, this, true,
+                     &m_routeChangeHandle);
+}
+
+WindowsBypass::~WindowsBypass() {
+  CancelMibChangeNotify2(m_netChangeHandle);
```

Comment:
This DTOR is never called. If yes, I wonder what guarantees that this object is correctly destroyed while the callbacks could be executed on any possible thread.

Reply:
The `WindowsBypass` class sets the `Socks5` as its parent object, meaning that their lifecycles are tied together. The DTOR will be called when the parent goes out of scope.

Also note, that `QObjects` without an explicit parent set are created as children of `qApp` and get destructed when the application exits.

If for some reason none of these DTORs run before the process quits, then it falls to the kernel cleanup misbehaving processes and safely de-register the callbacks.